README.md
# Domoticz Plugin homematicIP Pluggable Switch and Meter (HMIP-PSM)
Domoticz plugin for the homematicIP Pluggable Switch and Meter (HMIP-PSM).

# Objectives
* To switch the device on and off.
* To measure, in regular intervals, the power (W), energy (Wh), voltage (V), current (mA).

## Solution
The HMIP-PSM is connected to a homematic IP system.
The homematic IP system used is a [RaspberryMatic](https://raspberrymatic.de/) operating system running the Homematic Central-Control-Unit (CCU).
The CCU has the additional software XML-API CCU Addon installed.
Communication between Domoticz and the CCU is via HTTP XML-API requests with HTTP XML response.

In Domoticz, following device(s) is/are created:
(Type,SubType) [XML-API Device Datapoint Type] - Note)
* Energy (kWh) [ENERGY_COUNTER] [POWER] - Power is not used for a device, just logged only
* Voltage (Voltage) [VOLTAGE]
* Current (Current) [CURRENT]
* Switch (Switch) [STATE]

The device state is updated every 60 seconds (default).

If required, further actions can defined, by for example creating a dzVents script to send a notification / email (see below "dzVents Example").

## Hardware
Hardware subject to change.
* Raspberry Pi 3B+ (RaspberryMatic System)
* homematicIP Pluggable Switch and Meter (HMIP-PSM)

## Software
Software versions subject to change.
* Raspberry Pi OS ( Raspbian GNU/Linux 10 buster, kernel 5.4.83-v7l+)
* Domoticz 2020.2 (build 12847)
* RaspberryMatic 3.55.5.20201226 [info](https://raspberrymatic.de/)
* XML-API CCU Addon 1.20 [info](https://github.com/jens-maus/XML-API)
* Python 3.7.3
* Python module ElementTree

**Note on the Python Module ElementTree**
The Python Module **ElementTree XML API** is used to parse the XML-API response.
This module is part of the standard package and provides limited support for XPath expressions for locating elements in a tree. 

_Hint_
(Optional)
For full XPath support install the module **ElementPath** from the terminal command-line for Python 2.x and 3.x via pip:
``` 
sudo pip install elementpath
sudo pip3 install elementpath
```

## RaspberryMatic Prepare
The RaspberryMatic system has been setup according [these](https://github.com/jens-maus/RaspberryMatic) guidelines.

The XML-API CCU Addon is required and installed via the HomeMatic WebUI > Settings > Control panel > Additional software (download the latest version from previous URL shared).
**IMPORTANT**
Be aware of the security risk, in case the HomeMatic Control Center can be reached via the Internet without special protection (see XML-API Guidelines).

### XML-API Scripts
The XML-API provides various tool scripts, i.e. devices state list, device state or set new value and many more.
The scripts are submitted via HTTP XML-API requests.
The plugin makes selective use of scripts with device id and datapoint id's.
The device id is required to get the state of the device datapoints. The datapoint id's are required to get the state/value of device attributes.

#### Device ID (statelist.cgi)
Get Device ID (attribute "ise_id") from list of all devices with channels and current values: http://ccu-ip-address/addons/xmlapi/statelist.cgi.

From the HTTP XML-API response, the Device ID ("ise_id") is selected by searching for the
* Device Name (i.e. Schalt-Mess-Steckdose MakeLab) or 
* Device Channel (HMIP-PSM 0001D3C99C6AB3:3 Switch actuator). 
The data is obtained from the HomeMatic WebUI Home page > Status and control > Devices.
The Device "ise_id" is required for the plugin parameter _Mode1_.
HTTP XML-API response: Device ID = 1418.
```
...
<device name="Schalt-Mess-Steckdose MakeLab" ise_id="1418" unreach="false" config_pending="false">
	<channel...>
</device>
...
```
This script 
* has to be run once from a browser prior installing the plugin to get the device id as required by the plugin parameter Device ID ("mode1") and the next script.
* is not used in the plugin.

#### Channel Datapoint(s) (state.cgi)
Request the Channel Datapoint(s) for a Device ID to get value(s) for selected attribute(s): http://ccu-ip-address/addons/xmlapi/state.cgi?device_id=DEVICE_ISE_ID
The **Device ID 1418** is used as parameter to get the device state from which the attributes can be selected. 
The HTTP XML-API response lists eight Channels from which channels **HmIP-SWDM 001558A99D5A78:3** and **HMIP-PSM 0001D3C99C6AB3:6** are used.
The datapoint(s) used:
* Channel 3: type="STATE" with ise_id="1451" to set/get the state of the switch. The value is from valuetype 2, i.e. true (ON) or false (OFF).
* Channel 6: type=ENERGY_COUNTER, POWER, VOLTAGE, CURRENT.
The datapoint "ise_id" is required for the plugin parameter _Mode2_.
```
<device name="Schalt-Mess-Steckdose MakeLab" ise_id="1418" unreach="false" config_pending="false">
<channel name="Schalt-Mess-Steckdose MakeLab:0" ise_id="1419" index="0" visible="true" operate="false">
	<datapoint name="HmIP-RF.0001D3C99C6AB3:0.CONFIG_PENDING" type="CONFIG_PENDING" ise_id="1420" value="false" valuetype="2" valueunit="" timestamp="1611569595" operations="5"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:0.DUTY_CYCLE" type="DUTY_CYCLE" ise_id="1424" value="false" valuetype="2" valueunit="" timestamp="1611569595" operations="5"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:0.OPERATING_VOLTAGE" type="OPERATING_VOLTAGE" ise_id="1426" value="0.000000" valuetype="4" valueunit="V" timestamp="0" operations="5"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:0.OPERATING_VOLTAGE_STATUS" type="OPERATING_VOLTAGE_STATUS" ise_id="1427" value="0" valuetype="16" valueunit="" timestamp="0" operations="5"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:0.RSSI_DEVICE" type="RSSI_DEVICE" ise_id="1428" value="190" valuetype="8" valueunit="" timestamp="1611569595" operations="5"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:0.RSSI_PEER" type="RSSI_PEER" ise_id="1429" value="178" valuetype="8" valueunit="" timestamp="1611503042" operations="5"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:0.UNREACH" type="UNREACH" ise_id="1430" value="false" valuetype="2" valueunit="" timestamp="1611569595" operations="5"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:0.UPDATE_PENDING" type="UPDATE_PENDING" ise_id="1434" value="false" valuetype="2" valueunit="" timestamp="1610968122" operations="5"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:0.ACTUAL_TEMPERATURE" type="ACTUAL_TEMPERATURE" ise_id="3065" value="26.000000" valuetype="4" valueunit="°C" timestamp="1611569595" operations="5"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:0.ACTUAL_TEMPERATURE_STATUS" type="ACTUAL_TEMPERATURE_STATUS" ise_id="3066" value="0" valuetype="16" valueunit="" timestamp="1611569595" operations="5"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:0.ERROR_CODE" type="ERROR_CODE" ise_id="3067" value="0" valuetype="8" valueunit="" timestamp="1611569595" operations="5"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:0.ERROR_OVERHEAT" type="ERROR_OVERHEAT" ise_id="3068" value="false" valuetype="2" valueunit="" timestamp="1611569595" operations="5"/>
</channel>
	<channel name="HMIP-PSM 0001D3C99C6AB3:1" ise_id="1438" index="1" visible="true" operate="false">
	<datapoint name="HmIP-RF.0001D3C99C6AB3:1.PRESS_LONG" type="PRESS_LONG" ise_id="1439" value="" valuetype="2" valueunit="" timestamp="0" operations="4"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:1.PRESS_SHORT" type="PRESS_SHORT" ise_id="1440" value="" valuetype="2" valueunit="" timestamp="0" operations="4"/>
</channel>
<channel name="HMIP-PSM 0001D3C99C6AB3:2" ise_id="1441" index="2" visible="true" operate="false">
	<datapoint name="HmIP-RF.0001D3C99C6AB3:2.PROCESS" type="PROCESS" ise_id="1442" value="0" valuetype="16" valueunit="" timestamp="1611569595" operations="5"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:2.SECTION" type="SECTION" ise_id="1443" value="2" valuetype="16" valueunit="" timestamp="1611569595" operations="5"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:2.SECTION_STATUS" type="SECTION_STATUS" ise_id="1444" value="0" valuetype="16" valueunit="" timestamp="1611569595" operations="5"/>
<datapoint name="HmIP-RF.0001D3C99C6AB3:2.STATE" type="STATE" ise_id="1445" value="true" valuetype="2" valueunit="" timestamp="1611569595" operations="5"/>
</channel>
	<channel name="HMIP-PSM 0001D3C99C6AB3:3" ise_id="1446" index="3" visible="true" operate="true">
	<datapoint name="HmIP-RF.0001D3C99C6AB3:3.PROCESS" type="PROCESS" ise_id="1448" value="0" valuetype="16" valueunit="" timestamp="1611569595" operations="5"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:3.SECTION" type="SECTION" ise_id="1449" value="2" valuetype="16" valueunit="" timestamp="1611569595" operations="5"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:3.SECTION_STATUS" type="SECTION_STATUS" ise_id="1450" value="0" valuetype="16" valueunit="" timestamp="1611569595" operations="5"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:3.STATE" type="STATE" ise_id="1451" value="true" valuetype="2" valueunit="" timestamp="1611569595" operations="7"/>
<datapoint name="HmIP-RF.0001D3C99C6AB3:3.COMBINED_PARAMETER" type="COMBINED_PARAMETER" ise_id="9533" value="" valuetype="20" valueunit="" timestamp="0" operations="2"/>
</channel>
<channel name="HMIP-PSM 0001D3C99C6AB3:4" ise_id="1452" index="4" visible="true" operate="true">
	<datapoint name="HmIP-RF.0001D3C99C6AB3:4.PROCESS" type="PROCESS" ise_id="1454" value="0" valuetype="16" valueunit="" timestamp="1611569595" operations="5"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:4.SECTION" type="SECTION" ise_id="1455" value="0" valuetype="16" valueunit="" timestamp="1611569595" operations="5"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:4.SECTION_STATUS" type="SECTION_STATUS" ise_id="1456" value="0" valuetype="16" valueunit="" timestamp="1611569595" operations="5"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:4.STATE" type="STATE" ise_id="1457" value="false" valuetype="2" valueunit="" timestamp="1611569595" operations="7"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:4.COMBINED_PARAMETER" type="COMBINED_PARAMETER" ise_id="9534" value="" valuetype="20" valueunit="" timestamp="0" operations="2"/>
</channel>
	<channel name="HMIP-PSM 0001D3C99C6AB3:5" ise_id="1458" index="5" visible="true" operate="true">
	<datapoint name="HmIP-RF.0001D3C99C6AB3:5.PROCESS" type="PROCESS" ise_id="1460" value="0" valuetype="16" valueunit="" timestamp="1611569595" operations="5"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:5.SECTION" type="SECTION" ise_id="1461" value="0" valuetype="16" valueunit="" timestamp="1611569595" operations="5"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:5.SECTION_STATUS" type="SECTION_STATUS" ise_id="1462" value="0" valuetype="16" valueunit="" timestamp="1611569595" operations="5"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:5.STATE" type="STATE" ise_id="1463" value="false" valuetype="2" valueunit="" timestamp="1611569595" operations="7"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:5.COMBINED_PARAMETER" type="COMBINED_PARAMETER" ise_id="9535" value="" valuetype="20" valueunit="" timestamp="0" operations="2"/>
</channel>
<channel name="HMIP-PSM 0001D3C99C6AB3:6" ise_id="1464" index="6" visible="true" operate="false">
	<datapoint name="HmIP-RF.0001D3C99C6AB3:6.CURRENT" type="CURRENT" ise_id="1465" value="290.000000" valuetype="4" valueunit="mA" timestamp="1611569595" operations="5"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:6.CURRENT_STATUS" type="CURRENT_STATUS" ise_id="1466" value="0" valuetype="16" valueunit="" timestamp="1611569595" operations="5"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:6.ENERGY_COUNTER" type="ENERGY_COUNTER" ise_id="1467" value="17602.100000" valuetype="4" valueunit="Wh" timestamp="1611569595" operations="5"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:6.ENERGY_COUNTER_OVERFLOW" type="ENERGY_COUNTER_OVERFLOW" ise_id="1468" value="false" valuetype="2" valueunit="" timestamp="1611569595" operations="5"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:6.FREQUENCY" type="FREQUENCY" ise_id="1469" value="49.970000" valuetype="4" valueunit="Hz" timestamp="1611569595" operations="5"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:6.FREQUENCY_STATUS" type="FREQUENCY_STATUS" ise_id="1470" value="0" valuetype="16" valueunit="" timestamp="1611569595" operations="5"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:6.POWER" type="POWER" ise_id="1471" value="50.800000" valuetype="4" valueunit="W" timestamp="1611569595" operations="5"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:6.POWER_STATUS" type="POWER_STATUS" ise_id="1472" value="0" valuetype="16" valueunit="" timestamp="1611569595" operations="5"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:6.VOLTAGE" type="VOLTAGE" ise_id="1473" value="226.000000" valuetype="4" valueunit="V" timestamp="1611569595" operations="5"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:6.VOLTAGE_STATUS" type="VOLTAGE_STATUS" ise_id="1474" value="0" valuetype="16" valueunit="" timestamp="1611569595" operations="5"/>
</channel>
<channel name="HMIP-PSM 0001D3C99C6AB3:7" ise_id="1475" index="7" visible="true" operate="false"/>
<channel name="HMIP-PSM 0001D3C99C6AB3:8" ise_id="3069" index="8" visible="true" operate="true">
	<datapoint name="HmIP-RF.0001D3C99C6AB3:8.WEEK_PROGRAM_CHANNEL_LOCKS" type="WEEK_PROGRAM_CHANNEL_LOCKS" ise_id="3070" value="0" valuetype="16" valueunit="" timestamp="1611569595" operations="5"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:8.WEEK_PROGRAM_TARGET_CHANNEL_LOCK" type="WEEK_PROGRAM_TARGET_CHANNEL_LOCK" ise_id="3071" value="" valuetype="16" valueunit="" timestamp="0" operations="2"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:8.WEEK_PROGRAM_TARGET_CHANNEL_LOCKS" type="WEEK_PROGRAM_TARGET_CHANNEL_LOCKS" ise_id="3072" value="" valuetype="16" valueunit="" timestamp="0" operations="2"/>
	<datapoint name="HmIP-RF.0001D3C99C6AB3:8.COMBINED_PARAMETER" type="COMBINED_PARAMETER" ise_id="9536" value="" valuetype="20" valueunit="" timestamp="0" operations="2"/>
</channel>
</device>
```
This script 

CHANGELOG.md
# Changelog domoticz-plugin-hmip-psm

### v1.5.1 (Build 20210125)
* NEW: Changed datapoint attribute from ise_id to type. No need to define the datapoint ise_id anymore, only the device ise_id. 
* UPD: Minor changes.

### v1.1.0 (Build 20210118)
* CHG: Changed XML Parser from lxml to ElementTree (in Python standard package) - encountered lxml issue ("ImportError") adding multiple devices with Domoticz 2020 or higher (not found an lxml solution).
* UPD: Various minor improvements

### v1.0.3 (Build 20191223)
* UPD: various minor corrections.

### v1.0.1 (Build 20190706)
* UPD: Plugin definition enhanced; Changed error messages do domoticz.error.

### v1.0.0 (Build 20190702)
* NEW: First version published.


plugin.py
# Domoticz Home Automation - homematicIP Pluggable Switch and Meter (HMIP-PSM)
# Switch and measure, in regular intervals, the power (W), energy (Wh), voltage (V), current (mA).
# Dependencies:
# RaspberryMatic XML-API CCU Addon (https://github.com/hobbyquaker/XML-API)
# Library ElementTree (https://docs.python.org/3/library/xml.etree.elementtree.html#)
# Notes:
# 1. After every change: delete the hardware using the plugin homematicIP Thermostat
# 2. After every change: restart domoticz by running from terminal the command: sudo service domoticz.sh restart
# 3. Domoticz Python Plugin Development Documentation (https://www.domoticz.com/wiki/Developing_a_Python_plugin)
# 4. Only two adevice attributes are used. The plugin is flexible to add more attributes as required (examples WTH-2: HUMIDITY, eTRV-2: LEVEL)
#
# Author: Robert W.B. Linn
# Version: See plugin xml definition

"""
<plugin key="HmIP-PSM" name="homematicIP Pluggable Switch and Meter (HmIP-PSM)" author="rwbL" version="1.5.1 (Build 20210125)">
    <description>
        <h2>homematicIP Pluggable Switch and Meter (HmIP-PSM) v1.5.1</h2>
        <ul style="list-style-type:square">
            <li>Switch the device On or Off.</li>
            <li>Measure, in regular intervals, the power (W), energy (Wh), voltage (V), current (mA).</li>
        </ul>
        <h3>Domoticz Devices (Type,SubType) [XML-API Device Datapoint Type]</h3>
        <ul style="list-style-type:square">
            <li>Energy (kWh) [ENERGY_COUNTER] [POWER]</li>
            <li>Voltage (Voltage) [VOLTAGE]</li>
            <li>Current (Current) [CURRENT]</li>
            <li>Switch (Switch) [STATE]</li>
        </ul>
        <h3>Hardware Configuration</h3>
        <ul style="list-style-type:square">
            <li>CCU IP Address (default: 192.168.1.225)</li>
            <li>Device ID (default: 1418, get via XML-API script http://ccu-ip-address/addons/xmlapi/statelist.cgi)</li>
            <li>Datapoint ID STATE (default: 1451, get via XML-API script http://ccu-ip-address/addons/xmlapi/statelist.cgi using HomeMatic WebUI Device Channel, i.e. HmIP-PSM 0001D3C99C6AB3:3 (switch))</li>
        </ul>
    </description>
    <params>
        <param field="Address" label="CCU IP" width="200px" required="true" default="192.168.1.225"/>
        <param field="Mode1" label="Device ID" width="75px" required="true" default="1418"/>
        <param field="Mode2" label="Datapoint ID STATE" width="75px" required="true" default="1451"/>
        <param field="Mode5" label="Check Interval (sec)" width="75px" required="true" default="60"/>
        <param field="Mode6" label="Debug" width="75px">
            <options>
                <option label="True" value="Debug" default="true"/>
                <option label="False" value="Normal"/>
            </options>
        </param>
    </params>
</plugin>
"""

# Set the plugin version
PLUGINVERSION = "v1.5.1"
PLUGINSHORTDESCRIPTON = "HMIP-PSM"

## Imports
import Domoticz
import urllib
import urllib.request
from datetime import datetime
import json
import xml.etree.ElementTree as etree

## Domoticz device units used for creating & updating devices
UNIT_ENERGY_COUNTER = 1                 # TypeName: kWh
TYPE_ENERGY_COUNTER = "ENERGY_COUNTER"
TYPE_POWER = "POWER"                    # NOT USED = Measured only
UNIT_VOLTAGE = 2                        # TypeName: Voltage
TYPE_VOLTAGE = "VOLTAGE"
UNIT_CURRENT = 3                        # TypeName: Current (Single)
TYPE_CURRENT = "CURRENT"
UNIT_SWITCH = 4                         # TypeName: Switch
TYPE_STATE = "STATE"

# Task to perform
TASKSWITCH = 1
TASKMETER = 2

class BasePlugin:

    def __init__(self):

        # HTTP Connection
        self.httpConn = None
        self.httpConnected = 0

        # Task to complete - default is measure energy
        self.Task = TASKMETER

        # Switch state as string as the datapoint value defined 
        # The switchid is assigned during first check interval
        self.SwitchID = ""
        self.SwitchState = "unknown"
        
        # The Domoticz heartbeat is set to every 60 seconds. Do not use a higher value as Domoticz message "Error: hardware (N) thread seems to have ended unexpectedly"
        # The Soil Moisture Monitor is read every Parameter.Mode5 seconds. This is determined by using a hearbeatcounter which is triggered by:
        # (self.HeartbeatCounter * self.HeartbeatInterval) % int(Parameter.Mode5) = 0
        self.HeartbeatInterval = 60
        self.HeartbeatCounter = 0
        return

    def onStart(self):
        Domoticz.Debug(PLUGINSHORTDESCRIPTON + " " + PLUGINVERSION)
        Domoticz.Debug("onStart called")
        Domoticz.Debug("Debug Mode:" + Parameters["Mode6"])

        if Parameters["Mode6"] == "Debug":
            self.debug = True
            Domoticz.Debugging(1)
            DumpConfigToLog()

        # if there no  devices, create these
        if (len(Devices) == 0):
            Domoticz.Debug("Creating new devices ...")
            Domoticz.Device(Name="Energy", Unit=UNIT_ENERGY_COUNTER, TypeName="kWh", Used=1).Create()
            Domoticz.Debug("Device created: "+Devices[UNIT_ENERGY_COUNTER].Name)
            Domoticz.Device(Name="Voltage", Unit=UNIT_VOLTAGE, TypeName="Voltage", Used=1).Create()
            Domoticz.Debug("Device created: "+Devices[UNIT_VOLTAGE].Name)
            Domoticz.Device(Name="Current", Unit=UNIT_CURRENT, TypeName="Current (Single)", Used=1).Create()
            Domoticz.Debug("Device created: "+Devices[UNIT_CURRENT].Name)
            Domoticz.Device(Name="Powerswitch", Unit=UNIT_SWITCH, TypeName="Switch", Used=1).Create()
            Domoticz.Debug("Device created: "+Devices[UNIT_SWITCH].Name)
            Domoticz.Debug("Creating new devices: OK")

        # Set the switchid from parameter mode2
        self.SwitchID = Parameters["Mode2"]
        Domoticz.Debug("SwitchID: "+self.SwitchID)
        
        # Heartbeat
        Domoticz.Debug("Heartbeat set: "+Parameters["Mode5"])
        Domoticz.Heartbeat(self.HeartbeatInterval)
        return

    def onStop(self):
        Domoticz.Debug("Plugin is stopping.")

    # Send the url parameter (GET request)
    # If task = energy then to obtain device state information in xml format
    # If task = switch then switch using the self.switchstate
    # The http response is parsed in onMessage()
    def onConnect(self, Connection, Status, Description):
        Domoticz.Debug("onConnect called. Status: " + str(Status) + ", Description:" + Description)
        if (Status == 0):
            Domoticz.Debug("CCU connected successfully.")
            self.httpConnected = 1

            # request device id datapoints to get the energy data
            if self.Task == TASKMETER:
                ## url example = 'http://192.168.1.225/addons/xmlapi/state.cgi?device_id=1418'
                url = '/addons/xmlapi/state.cgi?device_id=' + Parameters["Mode1"]
                       
            # request state change for the switch with datapoint and new value
            if self.Task == TASKSWITCH:
                ## url example = 'http://192.168.1.225/addons/xmlapi/statechange.cgi?ise_id=1451&new_value=true'
                url = '/addons/xmlapi/statechange.cgi?ise_id=' + self.SwitchID + '&new_value=' + self.SwitchState

            Domoticz.Debug(url)
            sendData = {'Verb' : 'GET',
                        'URL' : url,
                        'Headers' : { 'Content-Type': 'text/xml; charset=utf-8', \
                                        'Connection': 'keep-alive', \
                                        'Accept': 'Content-Type: text/html; charset=UTF-8', \
                                        'Host': Parameters["Address"], \
                                        'User-Agent':'Domoticz/1.0' }
                       }
            self.httpConn.Send(sendData)
            self.httpConn.Disconnect
            return
        else:
            self.httpConnected = 0
            Domoticz.Error("IP connection faillure ("+str(Status)+") to: "+Parameters["Address"]+":"+Parameters["Port"]+" with error: "+Description)
            return

    # Parse the http xml response and update the domoticz devices
    def onMessage(self, Connection, Data):
        Domoticz.Debug("onMessage called")
        
        # If not conected, then leave
        if self.httpConnected == 0:
            return

        # Parse the JSON Data Object with keys Status (Number) and Data (ByteArray)
        ## 200 is OK
        responseStatus = int(Data["Status"])
        Domoticz.Debug("STATUS=responseStatus:" + str(responseStatus) + " ;Data[Status]="+Data["Status"])

        ## decode the data using the encoding as given in the xml response string
        responseData = Data["Data"].decode('ISO-8859-1')
        ## Domoticz.Debug("DATA=" + responseData)

        if (responseStatus != 200):
            Domoticz.Error("XML-API response faillure: " + str(responseStatus) + ";" + resonseData)
            return

        # Parse the xml string 
        # Get the xml tree - requires several conversions
        tree = etree.fromstring(bytes(responseData, encoding='utf-8'))

        # Handle the respective task to update the domoticz devices
        
        if self.Task == TASKMETER:
            Domoticz.Debug("TASKMETER")
            # Get the values for energy counter, power, voltage, current
            # <datapoint name="HmIP-RF.0001D3C99C6AB3:6.ENERGY_COUNTER" type="ENERGY_COUNTER" ise_id="1467" value="17221.300000" valuetype="4" valueunit="Wh" timestamp="1611423150"/>
            energycountervalue = tree.find(".//datapoint[@type='" + TYPE_ENERGY_COUNTER + "']").attrib['value']   # Domoticz.Debug(energycountervalue[0])
            #<datapoint name="HmIP-RF.0001D3C99C6AB3:6.POWER" type="POWER" ise_id="1471" value="49.890000" valuetype="4" valueunit="W" timestamp="1611423150"/>
            powervalue = tree.find(".//datapoint[@type='" + TYPE_POWER + "']").attrib['value']  # Domoticz.Debug(powervalue[0])
            # <datapoint name="HmIP-RF.0001D3C99C6AB3:6.VOLTAGE" type="VOLTAGE" ise_id="1473" value="224.200000" valuetype="4" valueunit="V" timestamp="1611423150"/>
            voltagevalue = tree.find(".//datapoint[@type='" + TYPE_VOLTAGE + "']").attrib['value']  # Domoticz.Debug(voltagevalue[0])
            # <datapoint name="HmIP-RF.0001D3C99C6AB3:6.CURRENT" type="CURRENT" ise_id="1465" value="285.000000" valuetype="4" valueunit="mA" timestamp="1611423150"/>
            currentvalue = tree.find(".//datapoint[@type='" + TYPE_CURRENT + "']").attrib['value']  # Domoticz.Debug(currentvalue[0])
            # <datapoint name="HmIP-RF.0001D3C99C6AB3:3.STATE" type="STATE" ise_id="1451" value="true" valuetype="2" valueunit="" timestamp="1611423150"/>
            switchstate = tree.find(".//datapoint[@type='" + TYPE_STATE + "']").attrib['value']                                       # Domoticz.Debug(switchstate[0])

            ## Update the devices and log
            Devices[UNIT_ENERGY_COUNTER].Update( nValue=0, sValue=str(round(float(powervalue),2)) + ";" + str(round(float(energycountervalue),2)) )
            Devices[UNIT_VOLTAGE].Update( nValue=0, sValue=str(round(float(voltagevalue),2)) )
            Devices[UNIT_CURRENT].Update( nValue=0, sValue=str(round(float(currentvalue) * 0.001,2)) )
            # Domoticz.Log(PLUGINSHORTDESCRIPTON + ": E=" + Devices[UNIT_ENERGY_COUNTER].sValue + ",V=" +  Devices[UNIT_VOLTAGE].sValue + ",A=" + Devices[UNIT_CURRENT].sValue + ",S=" + switchstate + ";" + str(Devices[UNIT_SWITCH].nValue) )

            ## Option to sync the switchstate if changed by the homematic webui
            if (switchstate == 'true') and (Devices[UNIT_SWITCH].nValue == 0):
                Devices[UNIT_SWITCH].Update( nValue=1, sValue= '0')
                Domoticz.Debug("Switch SYNC=" + switchstate + ";" + str(Devices[UNIT_SWITCH].nValue) )

            if (switchstate == 'false') and (Devices[UNIT_SWITCH].nValue == 1):
                Devices[UNIT_SWITCH].Update( nValue=0, sValue= '0')
                Domoticz.Debug("Switch SYNC=" + switchstate + ";" + str(Devices[UNIT_SWITCH].nValue) )

        if self.Task == TASKSWITCH:
            # The xml response data for the switch is like:
            # <?xml version="1.0" encoding="ISO-8859-1" ?><result><changed id="1451" new_value="false" /></result>
            Domoticz.Debug("TASKSWITCH:" + responseData)
            # Update the switch state
            if (self.SwitchState == 'true'):
                Devices[UNIT_SWITCH].Update( nValue=1, sValue= '0')
            else:
                Devices[UNIT_SWITCH].Update( nValue=0, sValue= '0')        

        return
        
    # Handle the state change of the switch. Convert state 'On' or 'Off' to 'true' or 'false' (as used by the xml-api script statecange)
    def onCommand(self, Unit, Command, Level, Hue):
        Domoticz.Debug("onCommand called for Unit " + str(Unit) + ": Parameter '" + str(Command) + "', Level: " + str(Level))
        
        if (Command == 'On'):
            self.SwitchState = 'true'
        else:
            self.SwitchState = 'false'

        # Create IP connection and connect - see further onConnect where the parameters are send
        self.httpConn = Domoticz.Connection(Name="CCU-"+Parameters["Address"], Transport="TCP/IP", Protocol="HTTP", Address=Parameters["Address"], Port="80")
        self.httpConn.Connect()
        self.httpConnected = 0
        self.Task = TASKSWITCH
        return

    def onNotification(self, Name, Subject, Text, Status, Priority, Sound, ImageFile):
        Domoticz.Debug("Notification: " + Name + "," + Subject + "," + Text + "," + Status + "," + str(Priority) + "," + Sound + "," + ImageFile)

    def onDisconnect(self, Connection):
        Domoticz.Debug("onDisconnect called")

    # Check the hearbeat counter and setup the http connection to the ccu
    # After connection, there url parameter are send - see onConnect()
    def onHeartbeat(self):
        self.HeartbeatCounter = self.HeartbeatCounter + 1
        Domoticz.Debug("onHeartbeat called. Counter=" + str(self.HeartbeatCounter * self.HeartbeatInterval) + " (Heartbeat=" + Parameters["Mode5"] + ")")
        # check the heartbeatcounter against the heartbeatinterval
        if (self.HeartbeatCounter * self.HeartbeatInterval) % int(Parameters["Mode5"]) == 0:
            try:
                # Create IP connection and connect
                self.httpConn = Domoticz.Connection(Name="CCU-"+Parameters["Address"], Transport="TCP/IP", Protocol="HTTP", Address=Parameters["Address"], Port="80")
                self.httpConn.Connect()
                self.httpConnected = 0
                self.Task = TASKMETER
            except:
                Domoticz.Error("IP connection faillure. Check settings and restart Domoticz.")
        return
        
global _plugin
_plugin = BasePlugin()

def onStart():
    global _plugin
    _plugin.onStart()

def onStop():
    global _plugin
    _plugin.onStop()

def onConnect(Connection, Status, Description):
    global _plugin
    _plugin.onConnect(Connection, Status, Description)

def onMessage(Connection, Data):
    global _plugin
    _plugin.onMessage(Connection, Data)

def onCommand(Unit, Command, Level, Hue):
    global _plugin
    _plugin.onCommand(Unit, Command, Level, Hue)

def onNotification(Name, Subject, Text, Status, Priority, Sound, ImageFile):
    global _plugin
    _plugin.onNotification(Name, Subject, Text, Status, Priority, Sound, ImageFile)

def onDisconnect(Connection):
    global _plugin
    _plugin.onDisconnect(Connection)

def onHeartbeat():
    global _plugin
    _plugin.onHeartbeat()

#
## Generic helper functions
#

def DumpConfigToLog():
    for x in Parameters:
        if Parameters[x] != "":
            Domoticz.Debug( "'" + x + "':'" + str(Parameters[x]) + "'")
    Domoticz.Debug("Device count: " + str(len(Devices)))
    for x in Devices:
        Domoticz.Debug("Device:           " + str(x) + " - " + str(Devices[x]))
        Domoticz.Debug("Device ID:       '" + str(Devices[x].ID) + "'")
        Domoticz.Debug("Device Name:     '" + Devices[x].Name + "'")
        Domoticz.Debug("Device nValue:    " + str(Devices[x].nValue))
        Domoticz.Debug("Device sValue:   '" + Devices[x].sValue + "'")
        Domoticz.Debug("Device LastLevel: " + str(Devices[x].LastLevel))
    return

